src/query_executor.c

```diff
@@ -84,143 +84,42 @@ static void _replicateMergeClauseToMatchClause(AST *ast) {
-     * 
-     * 2. Scanning though the temporary vector we expand each collapsed entity
-     * this will form the final RETURN clause. */
-
-    AST_ReturnNode *returnClause = ast->returnNode;
-    size_t returnElementCount = array_len(returnClause->returnElements);
-    AST_ReturnElementNode **elementsToExpand = array_new(AST_ReturnElementNode*, returnElementCount);
+/* If we have a "RETURN *" clause, populate it with all aliased entities. */
+static void _populateReturnAll(AST *ast) {
+    // Do nothing if there is no RETURN or an array of return elements already exists.
+    if (ast->returnNode == NULL || ast->returnNode->returnElements != NULL) return;
 
+    // Collect all entities from MATCH and CREATE clauses
+    // TODO Add entities from UNWIND
     TrieMap *identifiers = NewTrieMap();
     MatchClause_DefinedEntities(ast->matchNode, identifiers);
     CreateClause_DefinedEntities(ast->createNode, identifiers);
 
-    // Scan through return elements, see if we can find '*' marker.
-    for(int i = 0; i < returnElementCount; i++) {
-        AST_ReturnElementNode *retElement = returnClause->returnElements[i];        
-        if(retElement->asterisks) {
-            // TODO: '*' can not be mixed with any other expression!
-            array_clear(elementsToExpand);
-            /* Expand with "RETURN *" queries.
-             * Extract all entities from MATCH clause and add them to RETURN clause.
-             * These will get expended later on. */
-            char *ptr;
-            tm_len_t len;
-            void *value;
-            TrieMapIterator *it = TrieMap_Iterate(identifiers, "", 0);
-            while(TrieMapIterator_Next(it, &ptr, &len, &value)) {
-                AST_GraphEntity *entity = (AST_GraphEntity*)value;
-                if(entity->anonymous) continue;
-
-                len = MIN(255, len);
-                memcpy(buffer, ptr, len);
-                buffer[len] = '\0';
-
-                AST_ArithmeticExpressionNode *arNode = New_AST_AR_EXP_VariableOperandNode(buffer, NULL);
-                AST_ReturnElementNode *returnEntity = New_AST_ReturnElementNode(arNode, NULL);
-                elementsToExpand = array_append(elementsToExpand, returnEntity);
-            }
-
-            TrieMapIterator_Free(it);            
-            Free_AST_ReturnElementNode(retElement);
-            break;
-        }
-        else {
-            elementsToExpand = array_append(elementsToExpand, retElement);
-        }
+    // Allocate a new return element array to contain all user-provided aliases
+    AST_ReturnElementNode **entities = array_new(AST_ReturnElementNode*, identifiers->cardinality);
+    char *ptr;
+    tm_len_t len;
+    void *value;
+    TrieMapIterator *it = TrieMap_Iterate(identifiers, "", 0);
+    while(TrieMapIterator_Next(it, &ptr, &len, &value)) {
+        AST_GraphEntity *entity = value;
+        if(entity->anonymous) continue;
+
+        // Copy each alias string to the stack
+        char buffer[len + 1];
```

Comment:
We had previously allocated a `char buffer[256] outside the loop and then done a memcpy from `ptr` to that. We can go back to doing that, though I'd want to add an overflow check against `len`.

I don't see a way to not have a `memcpy` at all, since it's unsafe to write past `len` into the pointer populated by the trie.